pkg/controller/merged.go

```diff
@@ -1 +1,170 @@
 package controller
+
+import (
+	"maps"
+	"slices"
+	"sync"
+
+	"github.com/dharmab/skyeye/pkg/brevity"
+	"github.com/dharmab/skyeye/pkg/parser"
+	"github.com/dharmab/skyeye/pkg/spatial"
+	"github.com/dharmab/skyeye/pkg/trackfiles"
+	"github.com/rs/zerolog/log"
+)
+
+// mergeTracker tracks hostile IDs and the friendly IDs they have merged with.
+type mergeTracker struct {
+	merged map[uint64]map[uint64]struct{}
+	lock   sync.RWMutex
+}
+
+func newMergeTracker() *mergeTracker {
+	return &mergeTracker{
+		merged: make(map[uint64]map[uint64]struct{}),
+	}
+}
+
+// merge records that the given hostile has merged with the given friendly.
+func (t *mergeTracker) merge(hostileID, friendID uint64) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	friendIDs, ok := t.merged[hostileID]
+	if !ok {
+		friendIDs = make(map[uint64]struct{})
+		t.merged[hostileID] = friendIDs
+	}
+	friendIDs[friendID] = struct{}{}
+}
+
+// isMerged checks if the given hostile has merged with the given friendly.
+func (t *mergeTracker) isMerged(hostileID, friendID uint64) bool {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+	friendIDs, ok := t.merged[hostileID]
+	if !ok {
+		return false
+	}
+	_, ok = friendIDs[friendID]
+	return ok
+}
+
+// friendliesMergedWith returns the number of friendlies that the given hostile ID is merged with, if any.
+func (t *mergeTracker) friendliesMergedWith(hostileID uint64) int {
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+	friendIDs, ok := t.merged[hostileID]
+	if !ok {
+		return 0
+	}
+	return len(friendIDs)
+}
+
+// separate records that the given hostile and friendly IDs have exited the merge.
+func (t *mergeTracker) separate(hostileID, friendID uint64) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	friendIDs, ok := t.merged[hostileID]
+	if !ok {
+		return
+	}
+	delete(friendIDs, friendID)
+	if len(friendIDs) == 0 {
+		delete(t.merged, hostileID)
+	}
+}
+
+// remove removes the given ID from the merge tracker.
+func (t *mergeTracker) remove(id uint64) {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	_, ok := t.merged[id]
+	if ok {
+		delete(t.merged, id)
+	} else {
+		for hostileID, friendIDs := range t.merged {
+			delete(friendIDs, id)
+			if len(friendIDs) == 0 {
+				delete(t.merged, hostileID)
+			}
+		}
+	}
+}
+
+func (c *controller) broadcastMerges() {
+	merges := c.scope.Merges(c.coalition)
+	newMerges := make(map[brevity.Group][]*trackfiles.Trackfile)
+	for hostileGroup, friendlies := range merges {
+		newMergedFriendlies := map[uint64]*trackfiles.Trackfile{}
+		for _, hostileID := range hostileGroup.ObjectIDs() {
+			for _, friendly := range friendlies {
```

Comment:
(triple) nested for loops are generally a code smell that this should be split into different methods.

the whole broadcastMerges could probably be split up more into smaller logical chunks.

Reply:
I wish Go had simple syntax for a cross product 😭

Reply:
Factored out, this also reduced the cyclomatic complexity of the function